src/main/java/org/opensearch/geospatial/ip2geo/jobscheduler/DatasourceUpdateService.java

```diff
@@ -0,0 +1,209 @@
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.geospatial.ip2geo.jobscheduler;
+
+import java.io.IOException;
+import java.net.URL;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+
+import lombok.extern.log4j.Log4j2;
+
+import org.apache.commons.csv.CSVParser;
+import org.apache.commons.csv.CSVRecord;
+import org.opensearch.OpenSearchException;
+import org.opensearch.client.Client;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.settings.ClusterSettings;
+import org.opensearch.geospatial.ip2geo.common.DatasourceFacade;
+import org.opensearch.geospatial.ip2geo.common.DatasourceManifest;
+import org.opensearch.geospatial.ip2geo.common.DatasourceState;
+import org.opensearch.geospatial.ip2geo.common.GeoIpDataFacade;
+import org.opensearch.geospatial.ip2geo.common.Ip2GeoSettings;
+
+@Log4j2
+public class DatasourceUpdateService {
+    private final ClusterService clusterService;
+    private final ClusterSettings clusterSettings;
+    private final Client client;
+    private final DatasourceFacade datasourceFacade;
+    private final GeoIpDataFacade geoIpDataFacade;
+
+    public DatasourceUpdateService(
+        final ClusterService clusterService,
+        final Client client,
+        final DatasourceFacade datasourceFacade,
+        final GeoIpDataFacade geoIpDataFacade
+    ) {
+        this.clusterService = clusterService;
+        this.clusterSettings = clusterService.getClusterSettings();
+        this.client = client;
+        this.datasourceFacade = datasourceFacade;
+        this.geoIpDataFacade = geoIpDataFacade;
+    }
+
+    /**
+     * Update GeoIp data
+     *
+     * @param datasource
+     * @throws Exception
+     */
+    public void updateOrCreateGeoIpData(final Datasource datasource) throws Exception {
+        URL url = new URL(datasource.getEndpoint());
+        DatasourceManifest manifest = DatasourceManifest.Builder.build(url);
+
+        if (shouldUpdate(datasource, manifest) == false) {
+            log.info("Skipping GeoIP database update. Update is not required for {}", datasource.getId());
+            datasource.getUpdateStats().setLastSkippedAt(Instant.now());
+            datasourceFacade.updateDatasource(datasource);
+            return;
+        }
+
+        Instant startTime = Instant.now();
+        String indexName = setupIndex(manifest, datasource);
+        String[] header;
+        List<String> fieldsToStore;
+        try (CSVParser reader = geoIpDataFacade.getDatabaseReader(manifest)) {
+            header = validateAndReturnHeader(reader.iterator()).values();
+            fieldsToStore = Arrays.asList(header).subList(1, header.length);
+            if (!datasource.isCompatible(fieldsToStore)) {
```

Comment:
In some places i see x ==false is used and in another places !x is used. Shall we keep this consistent?